src/routing_core.rs

```diff
@@ -743,9 +743,313 @@ impl RoutingCore {
+                                        },
+                                    }
+                                },
+                                None => {},
+                            }
+                        }
+
+                        match opt_hello {
+                            Some(hello) => {
+                                let mut primary_connection = None;
+                                for (key, value) in self.unknown_connections.iter() {
+                                    match value.0 {
+                                        Some(ref value) => {
+                                            if *value == hello {
+                                                primary_connection = Some(key.clone());
+                                                break;
+                                            }
+                                        },
+                                        None => {},
+                                    }
+                                }
+                                match primary_connection {
+                                    Some(primary_connection) => {
+                                        // Try adding the peer to routing table.
+                                        if self.add_peer(ConnectionName::Routing(
+                                                hello.public_id.name()),  primary_connection,
+                                                hello.public_id.clone()) {
+                                            // Drop secondary, i.e., unrequired connection.
+                                            let _ = self.action_sender.send(
+                                                ::action::Action::DropConnections(
+                                                    vec![connection.clone()]));
+                                            self.remove_expected_connection(&expected_connection);
+                                        }
+                                    },
+                                    None => {},
+                                }
+                            },
+                            None => {},
+                        }
+                    }
+                }
+            },
+            (None, Some((unknown_connection, Some(hello)))) => {
+                // At matching from unknown_connection against expected connection
+                let mut opt_connection = None;
+                match hello.expected_connection {
+                    Some(ref hello_expected_connection) => {
+                        match *hello_expected_connection {
+                            ExpectedConnection::Request(ref request) => {
+                                for (key, value) in self.expected_connections.iter() {
+                                    match key {
+                                        &ExpectedConnection::Request(_) => {
+                                            // Don't match on ConnectRequest.
```

Comment:
I do believe we need to act here too.  We have a Hello and found a ConnectResponse